### PR TITLE
iron: upstream package updates

### PR DESCRIPTION
**Pull request by jimmy-mcelwain**
Updates packages to pull from newer versions for iron

### CODE REVIEW

**Comment by gavanderhoorn** [comment]
Thanks for the PR.

Could you please rebase on `iron_update_20240327`? Your PR contains a couple of changes that would already be part of #3 and the diff is a bit confusing like this.

**Comment by gavanderhoorn** [comment]
I'm especially confused about 1aa6c57559e16afb6fe9ec7b9f1a4424b83c08d6.

**Comment by jimmy-mcelwain** [comment]
Alright, I rebased onto `iron_update_20240327`. Hopefully that makes it cleaner.

File: target_ws_pkgs.ignore [hunk 1]
```diff
@@ -37,13 +37,16 @@ ros2/rmw_implementation/test_rmw_implementation
 ros2/rosidl/rosidl_generator_tests
 ros2/rosidl/rosidl_typesupport_introspection_cpp
 ros2/rosidl/rosidl_typesupport_introspection_tests
+ros2/rosidl/rosidl_generator_tests
```
Comment (gavanderhoorn) [inline target_ws_pkgs.ignore]:
```suggestion
```

not sure what happened here: is this a duplicate of line 37?

File: target_ws_pkgs.ignore [hunk 2]
```diff
@@ -37,13 +37,16 @@ ros2/rmw_implementation/test_rmw_implementation
...
 uros/rcl/rcl_lifecycle
 uros/rosidl_typesupport/rosidl_typesupport_tests
 uros/rosidl_typesupport_microxrcedds/test
+uros/rosidl_typesupport/rosidl_typesupport_tests
```
Comment (gavanderhoorn) [inline target_ws_pkgs.ignore]:
```suggestion
```

duplicate of line 42?

File: target_ws_pkgs.ignore [hunk 3]
```diff
@@ -37,13 +37,16 @@ ros2/rmw_implementation/test_rmw_implementation
...
 uros/tracetools/ros2trace
 uros/tracetools/test_ros2trace
 uros/tracetools/test_tracetools
 uros/tracetools/test_tracetools_launch
+uros/tracetools/test_ros2trace
```
Comment (gavanderhoorn) [inline target_ws_pkgs.ignore]:
```suggestion
```

duplicate of line 46?

File: target_ws_pkgs.repos [hunk 4]
```diff
@@ -39,9 +39,9 @@ repositories:
     url: https://github.com/ros2/rclc.git
     # NOTE: see the release notes: https://github.com/ros2/rclc/releases/tag/5.0.1
     # especially the PRs touching rosout logging init
-    # NOTE 2: this is two commit past '5.0.1', fixing the clash with 'NONE'.
+    # NOTE 2: this is three commits past '5.0.1', fixing the clash with 'NONE'.
     # See: https://github.com/ros2/rclc/issues/378
-    version: 24ab126bb489984c2b66a72d2ce56e221dd707d1
+    version: 63d5c43993cb2d0d03ecd1b68d771046fd4ad19f
```
Comment (gavanderhoorn) [inline target_ws_pkgs.repos]: I would suggest we keep this as it was. The extra commit only changes CI config in `rclc`: https://github.com/ros2/rclc/commit/63d5c43993cb2d0d03ecd1b68d771046fd4ad19f.

File: target_ws_pkgs.repos [hunk 5]
```diff
@@ -52,6 +52,8 @@ repositories:
   ros2/rmw:
     type: git
     url: https://github.com/yaskawa-global/rmw.git
+    # one patch on top of iron (which removes the message string
+    # from GCC deprecation attributes used (which GCC in M+ doesn't support))
```
Comment (gavanderhoorn) [inline target_ws_pkgs.repos]:
```suggestion
```

File: target_ws_pkgs.repos [hunk 6]
```diff
@@ -97,12 +99,11 @@ repositories:
...
     type: git
     url: https://github.com/micro-ROS/micro_ros_utilities.git
-    # three commits past 4.0.1
-    version: 74f7a6b08690079ecbb0262db563aa96f667277d
+    version: 4.0.2
   uros/rmw-microxrcedds:
     type: git
     url: https://github.com/micro-ROS/rmw-microxrcedds.git
-    version: 4.0.1
+    version: 4.0.2
```
Comment (gavanderhoorn) [inline target_ws_pkgs.repos]: Also suggest to revert this: `4.0.2` only updates Doxygen config and some minor content changes.

File: target_ws_pkgs.repos [hunk 7]
```diff
@@ -132,7 +133,8 @@ repositories:
     url: https://github.com/ros2/unique_identifier_msgs.git
     version: 2.3.2
 
-  # only used to determine the ROS 2 codename during packaging
+  # we only use this to determine the ROS 2 distro codename during packaging.
+  # It does not end up in 'libmicroros.a'
```
Comment (gavanderhoorn) [inline target_ws_pkgs.repos]:
```suggestion
```

File: target_ws_pkgs.repos [hunk 8]
```diff
@@ -147,12 +149,12 @@ repositories:
...
     url: https://github.com/ros-controls/control_msgs.git
-    # technically on 'rolling', but this seems the version released into Iron ..
-    version: 5.0.0
+    # technically on 'rolling', but this is the version released into Iron ..
+    version: 5.2.0
   extra/ros2/geometry2_tf2_msgs:
     type: git
     url: https://github.com/ros2/geometry2.git
-    version: 0.31.5
+    version: 0.31.7
```
Comment (gavanderhoorn) [inline target_ws_pkgs.repos]: we only use `tf2_msgs` from this, and both `0.31.6` and `0.31.7` don't contain any changes to that package ([here](https://github.com/ros2/geometry2/commit/719c3a212480f8f418ba682ba53c623687bb366d#diff-ca78c11f23eece3ef5739c290f99b57bff30c3d7c44e630056580cc52de3701e) and [here](https://github.com/ros2/geometry2/commit/c2b74336100e5a24720dccaac102f181511bd7dd#diff-ca78c11f23eece3ef5739c290f99b57bff30c3d7c44e630056580cc52de3701e)).

**Review by gavanderhoorn — CHANGES_REQUESTED** [review]

**Comment by gavanderhoorn** [comment]
Thanks @jimmy-mcelwain.

I've squashed your commits (and a single fixup from my side).

**Comment by gavanderhoorn** [comment]
Could you please let me know which upstream Iron release you took these package versions from?

Then we can update the commit comment (on 0ec37d6db619b93949288d148c33d32ffb57ad63) and merge this.

**Comment by jimmy-mcelwain** [comment]
I misunderstood, I didn't take the versions from any single release of iron. In many cases I just pulled the latest release for iron from each package individually. I can update it to match a single released version (I will look at [20240423](https://github.com/ros2/ros2/releases/tag/release-iron-20240423)), but to do so I may have to also rebase some of our forked branches onto newer upstream versions, e.g. our fork of [ament_cmake](https://github.com/Yaskawa-Global/ament_cmake/tree/iron-motoplus1-candidate-20240327) is based on 2.0.4 whereas iron release 20240423 targets uses 2.0.5. 

Also, the [micro-ros releases](https://github.com/micro-ROS/micro_ros_setup/releases/tag/4.1.0) don't appear to target specific releases/commits when pulling packages, it just pulls the newest version from each iron branch [if I am reading the .repos file correctly](https://github.com/micro-ROS/micro_ros_setup/blob/iron/config/client_uros_packages.repos). So I may just target the latest released version for each repo that they target if that seems reasonable.

Sorry for the confusion

**Comment by gavanderhoorn** [comment]
First: lets not make things more complex than they already are. I suggest we keep the list of pkgs you already identified, including the versions in this PR.

Having written that: we loosely follow [calendar versioning](https://calver.org) here, and the fact upstream tags their `.repos` files with something like that should make things easier to keep in-sync. Rationale: (expected) benefits wrt compatibility between (micro)ROS 2 and MotoROS2.

We can't completely limit ourselves to upstream releases as we sometimes need additional (unreleased) fixes.

For future updates we'll want to look at / stick to upstream release tags. For now we'll keep your PR as-is.

**Comment by gavanderhoorn** [comment]
I suggest we update the commit comment to instead mention the date at which you checked upstream.

**Comment by gavanderhoorn** [comment]
oh, and to clarify:

> Also, the [micro-ros releases](https://github.com/micro-ROS/micro_ros_setup/releases/tag/4.1.0) don't appear to target specific releases/commits when pulling packages, it just pulls the newest version from each iron branch [if I am reading the .repos file correctly](https://github.com/micro-ROS/micro_ros_setup/blob/iron/config/client_uros_packages.repos?rgh-link-date=2024-07-02T15%3A27%3A20Z). So I may just target the latest released version for each repo that they target if that seems reasonable.

When I write *upstream*, I don't necessarily mean micro-ROS. Technically we use micro-ROS, but quite a few packages come from 'regular' ROS 2.

I should've clarified.

**Comment by jimmy-mcelwain** [comment]
Okay. I last did a complete check of the upstream branches yesterday, so I marked it as such through the commit message and pushed it back up.

**Comment by gavanderhoorn** [comment]
Test build: [here](https://github.com/Yaskawa-Global/motoplus_libmicroros_builder/actions/runs/9857358630).

Seems to build. Thanks @jimmy-mcelwain :+1:

**Review by gavanderhoorn — APPROVED** [review]

**Pull request merged**